File: 26-sum/prog/td-game/src/game/map.cpp
#include "map.h"

Map::Map(float width, float height, std::vector<Tile> const& tiles)
    : width_(width), height_(height), tiles_(tiles) { }

float Map::width() const { return width_; }
float Map::height() const { return height_; }
std::vector<Tile> const& Map::tiles() const { return tiles_; }

void Map::clear_occupancy() const {
  for (auto const& tile : tiles_) {
    tile.set_occupied_by_tower(false);
  }
}

Tile const* Map::tile_at(Vec2 position) const {
  for (auto const& tile : tiles_) {
    if (tile.bounds().contains(position)) {
      return &tile;
    }
  }
  return nullptr;
}

File: 26-sum/prog/td-game/src/rust/src/map.rs
use crate::{geometry::Vec2, tile::Tile};

/// A map is a rectangular area that contains tiles, defining the playable area.
#[derive(Debug, Clone)]
pub struct Map {
    pub width: f32,
    pub height: f32,
    pub tiles: Vec<Tile>,
}

impl Map {
    pub fn new(width: f32, height: f32, tiles: Vec<Tile>) -> Self {
        Self {
            width,
            height,
            tiles,
        }
    }

    pub fn tile_at(&self, position: Vec2) -> Option<&Tile> {
        self.tiles
            .iter()
            .find(|tile| tile.bounds.contains(position))
    }

    pub fn tile_at_mut(&mut self, position: Vec2) -> Option<&mut Tile> {
        self.tiles
            .iter_mut()
            .find(|tile| tile.bounds.contains(position))
    }

    pub fn tile_with_index_at(&self, position: Vec2) -> Option<(usize, &Tile)> {
        self.tiles
            .iter()
            .enumerate()
            .find(|(_, tile)| tile.bounds.contains(position))
    }
}
